File: cs/algorithms_in/heap_sort.cs
namespace algorithms_in;

public class HeapSort
{
    static Heap heap = new Heap();

    public static void heapSort(int[] array) {
        for (int i = 0; i < array.Length; i++) {
            heap.add(array[i]);
        }

        for (int i = array.Length - 1; i >= 0; i--) {
            array[i] = heap.remove();
        }
    }
}

public class Heap
{
    ArrayList list;
    public Heap() {
        list = new ArrayList();
    }

    public void add(int value) {
        list.add(value);

        int currentIndex = list.size() - 1;

        while (currentIndex > 0) {
            int parentIndex = (currentIndex - 1) / 2;

            if (compare(currentIndex, parentIndex) > 0) {
                int temp = list.get(currentIndex);
                list.set(currentIndex, list.get(parentIndex));
                list.set(parentIndex, temp);
            } else {
                break;
            }

            currentIndex = parentIndex;
        }
    }

    public int remove() {
        int removedValue = list.get(0);

        list.set(0, list.get(list.size() - 1));
        list.remove();

        int currentIndex = 0;
        while (currentIndex < list.size()) {
            int leftChildIndex = 2 * currentIndex + 1;
            int rightChildIndex = 2 * currentIndex + 2;

            if (leftChildIndex >= list.size()) {
                break;
            }

            int maxIndex = leftChildIndex;
            if (rightChildIndex < list.size()) {
                if (compare(maxIndex, rightChildIndex) < 0) {
                    maxIndex = rightChildIndex;
                }
            }

            if (compare(currentIndex, maxIndex) < 0) {
                int temp = list.get(maxIndex);
                list.set(maxIndex, list.get(currentIndex));
                list.set(currentIndex, temp);
                currentIndex = maxIndex;
            } else {
                break;
            }
        }

        return removedValue;
    }

    private int compare(int v1, int v2) {
        return list.get(v1) - list.get(v2);
    }
}

public class ArrayList
{
    List<int> list;
    public ArrayList() {
        list = new List<int>();
    }

    public void add(int value) {
        list.Add(value);
    }

    public void remove() {
        list.RemoveAt(list.Count - 1);
    }

    public int get(int index) {
        return list[index];
    }

    public void set(int index, int value) {
        list[index] = value;
    }

    public int size() {
        return list.Count();
    }
}

File: go/heap_sort/heap_sort.go
package algorithms_in

type Heap struct {
	list ArrayList
}

func (h *Heap) add(value int) {
	h.list.add(value)

	currentIndex := h.list.size() - 1

	for currentIndex > 0 {
		parentIndex := (currentIndex - 1) / 2
		
		if h.compare(currentIndex, parentIndex) > 0 {
			temp := h.list.get(currentIndex)
			h.list.set(currentIndex, h.list.get(parentIndex))
			h.list.set(parentIndex, temp)
		} else {
			break
		}
		currentIndex = parentIndex
	}
}

func (h *Heap) remove() int {
	removedValue := h.list.get(0)

	h.list.set(0, h.list.get(h.list.size() - 1))
	h.list.remove()

	currentIndex := 0
	for currentIndex < h.list.size() {
		leftChildIndex := 2 * currentIndex + 1
		rightChildIndex := 2 * currentIndex + 2

		if leftChildIndex >= h.list.size() {
			break
		}

		maxIndex := leftChildIndex
		if rightChildIndex < h.list.size() {
			if h.compare(maxIndex, rightChildIndex) < 0 {
				maxIndex = rightChildIndex
			}
		}

		if h.compare(currentIndex, maxIndex) < 0 {
			temp := h.list.get(maxIndex)
			h.list.set(maxIndex, h.list.get(currentIndex))
			h.list.set(currentIndex, temp)
			currentIndex = maxIndex
		} else {
			break
		}
	}

	return removedValue
}

func (h *Heap) compare(v1, v2 int) int {
	return h.list.get(v1) - h.list.get(v2)
}

type ArrayList struct {
	list []int
}

func (a *ArrayList) add(value int) {
	a.list = append(a.list, value)
}

func (a *ArrayList) remove() {
	a.list = a.list[:len(a.list) - 1]
}

func (a *ArrayList) get(index int) int {
	return a.list[index]
}

func (a *ArrayList) set(index, value int) {
	a.list[index] = value
}

func (a *ArrayList) size() int {
	return len(a.list)
}

func HeapSort(list []int) {
	heap := Heap{}

	for i := range list {
		heap.add(list[i])
	}

	for i := len(list) - 1; i >= 0; i-- {
		list[i] = heap.remove()
	}
}